apps/speed.c

```diff
@@ -3161,6 +3566,369 @@ int speed_main(int argc, char **argv)
         }
     }
 #endif  /* OPENSSL_NO_DH */
+
+    for (testnum = 0; testnum < kems_algs_len; testnum++) {
+        int kem_checks = 1;
+        const char *kem_name = kems_algname[testnum];
+
+        if (!kems_doit[testnum] || !do_kems)
+            continue;
+
+        for (i = 0; i < loopargs_len; i++) {
+            EVP_PKEY *pkey = NULL;
+            EVP_PKEY_CTX *kem_gen_ctx = NULL;
+            EVP_PKEY_CTX *kem_encaps_ctx = NULL;
+            EVP_PKEY_CTX *kem_decaps_ctx = NULL;
+            size_t send_secret_len, out_len;
+            size_t rcv_secret_len;
+            unsigned char *out = NULL, *send_secret = NULL, *rcv_secret;
+            size_t bits;
+            char *name;
+            OSSL_PARAM params[] = { OSSL_PARAM_END, OSSL_PARAM_END };
+            int use_params = 0;
+
+            if (ERR_peek_error()) {
+                BIO_printf(bio_err,
+                           "WARNING: the error queue contains previous unhandled errors.\n");
+                ERR_print_errors(bio_err);
+            }
+
+
+            if (strncmp(kem_name, "RSA", 3) == 0) {
+                bits = atoi(kem_name+3);
+                params[0] = OSSL_PARAM_construct_size_t(OSSL_PKEY_PARAM_RSA_BITS,
+                                                        &bits);
+                use_params = 1;
+            } else if (strncmp(kem_name, "EC", 2) == 0) {
+                name = (char *)(kem_name+2);
+                params[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME,
+                                                  name, 0);
+                use_params = 1;
+            }
+
+            kem_gen_ctx = EVP_PKEY_CTX_new_from_name(app_get0_libctx(),
+                                               !strncmp(kem_name, "RSA", 3)?"RSA":
+                                                !strncmp(kem_name, "EC", 2)?"EC":
+                                                 kem_name,
+                                               app_get0_propq());
+
+            if (!kem_gen_ctx || EVP_PKEY_keygen_init(kem_gen_ctx) <= 0 ||
+                (use_params &&
+                 (EVP_PKEY_CTX_set_params(kem_gen_ctx, params) <= 0))) {
+                BIO_printf(bio_err, "Error initializing keygen ctx for %s.\n",
+                           kem_name);
+                goto kem_err_break;
+            }
+            if (EVP_PKEY_keygen(kem_gen_ctx, &pkey) <= 0) {
+                BIO_printf(bio_err, "Error while generating KEM EVP_PKEY.\n");
+                goto kem_err_break;
+            }
+            /* Now prepare encaps data structs */
+            kem_encaps_ctx = EVP_PKEY_CTX_new_from_pkey(app_get0_libctx(),
+                                                        pkey,
+                                                        app_get0_propq());
+            if ((kem_encaps_ctx == NULL) ||
+                (EVP_PKEY_encapsulate_init(kem_encaps_ctx, NULL) <= 0)  ||
+                ((!strncmp(kem_name, "RSA", 3) &&
```

Comment:
nit: explicit test against 0, please

Reply:
changed here and in other places in https://github.com/openssl/openssl/pull/19968/commits/b7b2a70ec374b2d556e1290a0abceb48009ae939